src/data/finsupp/lattice.lean

```diff
@@ -0,0 +1,119 @@
+/-
+Copyright (c) 2020 Aaron Anderson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Aaron Anderson
+-/
+import data.finsupp
+import tactic.squeeze
+import algebra.ordered_group
+--open finsupp
+
+/-!
+# Lattice structure on finsupps
+
+This file provides instances of ordered structures on finsupps.
+
+-/
+
+open_locale classical
+noncomputable theory
+variables {α : Type*} {β : Type*} [has_zero β] {μ : Type*} [canonically_ordered_add_monoid μ]
+variables {γ : Type*} [canonically_linear_ordered_add_monoid γ]
+
+namespace finsupp
+
+lemma le_def [partial_order β] {a b : α →₀ β} : a ≤ b ↔ ∀ (s : α), a s ≤ b s := by refl
+
+instance : order_bot (α →₀ μ) :=
+{ bot := 0, bot_le := by simp [finsupp.le_def, ← bot_eq_zero], .. finsupp.partial_order}
+
+/-- Used to construct binary operations on `finsupp`s -/
+def binary_op_pointwise {f : β → β → β} (h : f 0 0 = 0) (a b : α →₀ β) : α →₀ β :=
+{ support := ((a.support) ∪ (b.support)).filter (λ s, f (a s) (b s) ≠ 0),
+  to_fun := λ s, f (a s) (b s),
+  mem_support_to_fun := by {
+    simp only [mem_support_iff, ne.def, finset.mem_union, finset.mem_filter],
+    intro s, apply and_iff_right_of_imp, rw ← not_and_distrib, rw not_imp_not,
+    intro h1, rw h1.left, rw h1.right, apply h, }, }
+
+lemma binary_op_pointwise_apply {f : β → β → β} (h : f 0 0 = 0) (a b : α →₀ β) (s : α) :
+  (binary_op_pointwise h a b) s = f (a s) (b s) := rfl
```

Comment:
Move to `data.finsupp`?

Reply:
Actually, shouldn't it go to `data.finsupp.pointwise`?

Reply:
Oh I see, this is basically what `zip_with` is...

Reply:
I should see if I can refactor it all to use zip_with, but my oleans are broken again, and the command I used earlier to fix it didn't work, and I've apparently exceeded the API rate limit trying to get new oleans...